### ISSUE TITLE
Used up too much memory?

### ISSUE THREAD

**Issue by mike32jacobs** [susanBuck/e2-fall21#32]
I wrote the code for my game of war using a short deck. I finished the implementation successfully. Before I turned it in, I add the missing cards to the deck. My game stopped working and I got the following error.
"Fatal error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 33554440 bytes) in /var/www/e2/p1/index.php on line 357"

I know that I could be more efficient, and allocate less memory. This is one of the issues that I anticipated in my previous post  [P1-War: Should I store the results of each round in an array? Or, immediately display the results?](https://github.com/susanBuck/e2-fall21/issues/30)

I am interested in finding out if the error is about the way in which I created the deck of cards, or if it was with the way in which I stored the data for each round of the game. Do you have any insight that could help. 

I pushed the most recent version of my code onto github. I assume that you could access it. [Link to my project](https://github.com/mike32jacobs/e2/tree/main/p1)

**Comment by susanBuck** [comment]
I cloned a copy of your code to play around with it, and narrowed it down to the game running into an infinite loop.

To troubleshoot this, I put in a "stop" in the while loop so I could have the program run and output results before reaching the out of memory error. This was done using PHP's built in "die" function:

```php
if ($roundsPlayed > 500) {
        var_dump('Rounds exceeded round limit, stop.');
        die();
}
```

With this in place, I also threw in a bunch of debugging output in the while loop so I could study the results of each round:
```php
 # Debugging 
    echo "Round: $roundsPlayed <br>";
    echo "Player 1 card count: ".count($player1Hand)." <br>";
    echo "Player 2 card count: ".count($player2Hand)." <br>";
    echo "Player 1 card: $p1Card[2]<br>";
    echo "Player 2 card: $p2Card[2]<br>";
    echo "Winner: $roundWinner<br>";
```

What I found is that with the full deck cards, the game would reach a state where it just alternated back and forth between Player 1 and Player 2 winning the rounds, resulting in neither player winning the game.

[Here's the example output from a game...](https://gist.github.com/susanBuck/d08adbb495ae75779d2af5732d061d7c)

Notice at at round 183, we get stuck in the alternating loop of Player 1 wins, then Player 2 wins, then Player 1 wins, then Player 2 Wins, etc.

The result is their decks don't shrink, they just flip back and forth between 37/3 and 36/4. 

I was trying to figure out why this doesn't happen "in real life" when playing this game and realized there's a crucial step we're missing - shuffling. In the real life game, you have your pile you're discarding from, and the pile you're collecting to. Whenever  the pile you're discarding from runs out, you pick up the pile you're collecting to, shuffle it, and then that becomes the pile you're discarding from.

I think by not having this occasional shuffle, the game runs into a state where the same cards are played against another and it just keeps flipping back and forth.

As a simple solution, I put in a shuffle on each player's hand before they discard:

```php
shuffle($player1Hand);
shuffle($player2Hand);
$p1Card = array_shift($player1Hand);
$p2Card = array_shift($player2Hand);
```

And that seemed to do the trick - the game would always play out in less than 300 rounds, well within the memory limit.

On a related note, as FYI:
On our servers PHP's default memory limit is configured to 128MB by default. You were seeing the error you were because your script was exceeding this limit. The memory limit is set in the php.ini (`/etc/php/7.4/fpm/php.ini`) file on line 409. (Recall this is the same file we changed our debug settings.) Ref: https://www.php.net/manual/en/ini.core.php#ini.memory-limit

PHP's memory limit can be upped if you have memory intensive scripts, but shouldn't be upped just to deal with "run away" scripts like we saw here. Always best to get to the source of the problem in those situations.

**Comment by mike32jacobs** [comment]
Thank you so much for investigating. That all makes so much sense. I think that it worked with one suit, because there was an odd number of cards, so it won't end up in an endless loop. I checked this by running the same code using 3 of the 4 suits, and the program worked as intended. 

Before I started the game, I did wonder about the reshuffle, but did not consider that could be the cause of my problem. I will change the rules to make sure I include the rule about reshuffling.
